### PR TITLE
[SPARK-32002][SQL]Support ExtractValue from nested ArrayStruct

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?

For data `nest.json`

```json
{"a": [{"b": [{"c": [1,2]}]}]}
{"a": [{"b": [{"c": [1]}, {"c": [2]}]}]}
```
run with 

```scala
val df: DataFrame = spark.read.json(testFile("nest-data.json"))
df.createTempView("nest_table")
sql("select a.b.c from nest_table").show()
```

will got error 

```
org.apache.spark.sql.AnalysisException: cannot resolve 'nest_table.`a`.`b`['c']' due to data type mismatch: argument 2 requires integral type, however, ''c'' is of string type.; line 1 pos 7;
'Project [a#6.b[c] AS c#8|#6.b[c] AS c#8]
+- SubqueryAlias `nest_table`
+- Relationa#6 json
```

This is because after resolve `a.b`, it construct `GetArrayStructField` from `a#`
result with type `ArrayType(ArrayType(StructType))`, then it can't be extract through current `ExtractValue` method.

Support extract `StructField` form nested Array-Struct combination

### Why are the changes needed?

Support more use case 

### Does this PR introduce _any_ user-facing change?

Yes, users will be able to select from nested array and structs.

### How was this patch tested?

Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -3521,6 +3523,25 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
             |""".stripMargin), Row(1))
     }
   }
+
+  test("SPARK-32002: Support Extract valve from nested ArrayStruct") {
```
Comment (Tagar) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3527]:
```suggestion
  test("SPARK-32002: Support Extract value from nested ArrayStruct") {
```
typo

**Comment by gatorsmile** [comment]
cc @MaxGekk @HyukjinKwon

**Comment by AngersZhuuuu** [comment]
Seems this way will change the schema, maybe we should add a new class to get multidimensional array form GetArrayStruct

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
cc @viirya too FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 2]
```diff
@@ -59,6 +59,23 @@ object ExtractValue {
         GetArrayStructFields(child, fields(ordinal).copy(name = fieldName),
           ordinal, fields.length, containsNull)
 
+      case (ExtractNestedArray(StructType(fields), containsNull, containsNullSeq),
+      NonNullLiteral(v, StringType)) =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:63]: nit:

```
case (ExtractNestedArray(
    StructType(fields), containsNull, containsNullSeq), NonNullLiteral(v, StringType)) =>
```

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:63]: Let's also update the documentation and table above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 3]
```diff
@@ -95,6 +112,50 @@ object ExtractValue {
 
 trait ExtractValue extends Expression
 
+object ExtractNestedArray {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:115]: Let's add documentation here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 4]
```diff
@@ -95,6 +112,50 @@ object ExtractValue {
...
+
+  type ReturnType = Option[(DataType, Boolean, Seq[Boolean])]
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:117]: Let's also add some comments for what this type means.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 5]
```diff
@@ -95,6 +112,50 @@ object ExtractValue {
...
+
+  type ReturnType = Option[(Expression, Int)]
+
+  def unapply(expr: Expression): ReturnType = {
+    extractArrayStruct(expr)
+  }
+
+  def extractArrayStruct(expr: Expression): ReturnType = {
+    expr match {
+      case gas @ GetArrayStructFields(child, _, _, _, _) =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:148]: Let's avoid arguments matching. This is actually an anti pattern - https://github.com/databricks/scala-style-guide#pattern-matching

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 6]
```diff
@@ -95,6 +112,50 @@ object ExtractValue {
...
+        extractArrayStruct(child) match {
+          case Some((e, deep)) => Some(e, deep + 1)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:150]: `deep` -> `depth`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 7]
```diff
@@ -95,6 +112,50 @@ object ExtractValue {
...
 
+object ExtractNestedArray {
+
+  type ReturnType = Option[(DataType, Boolean, Seq[Boolean])]
+
+  def unapply(dataType: DataType): ReturnType = {
+     extractArrayType(dataType)
+  }
+
+  def extractArrayType(dataType: DataType): ReturnType = {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:123]: nit: `private`

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:123]: Can we combine this and `unapply`?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 8]
```diff
@@ -3521,6 +3523,25 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+      val df = spark.read
+        .json(Seq(
+          """{"a": [{"b": [{"c": [1,2]}]}]}""",
+          """{"a": [{"b": [{"c": [1]}, {"c": [2]}]}]}""",
+          """{"a":[{}]}""").toDS())
+      df.createOrReplaceTempView("nest")
+
+      checkAnswer(sql(
+        """
+          |SELECT a.b.c FROM nest
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3538]: Can we add deeper cases? Also, you can simplify the test cases, for example, as below:

```scala
val df = spark.range(10).select(array(struct(array(struct(id)).alias(col1))).alias(col0))
df.selectExpr("col0.col1.id")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 3 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:115]: Let's also name it something like `NestedArrayType`

**Comment by guiyanakuang** [comment]
ProjectionOverSchema.scala. A Scala extractor that projects an expression over a given schema.    I think ProjectionOverSchema.scala must add matching rules as well.

**Comment by guiyanakuang** [comment]
A Scala extractor that projects an expression over a given schema.

**Comment by guiyanakuang** [comment]
org.apache.spark.sql.execution.ProjectionOverSchema.scala. A Scala extractor that projects an expression over a given schema. I think this code also needs to have matching rules added to it. To support nested ArrayStruct. Otherwise, ParquetSchemaPruning won't work either.

**Comment by AngersZhuuuu** [comment]
> org.apache.spark.sql.execution.ProjectionOverSchema.scala. A Scala extractor that projects an expression over a given schema. I think this code also needs to have matching rules added to it. To support nested ArrayStruct. Otherwise, ParquetSchemaPruning won't work either.

Yea, right, current way change the schema and won't work well for all case, I am thinking about add a new class to handle nested ArrayStruct type

**Pull request closed without merging**